hack/generator/pkg/astmodel/package_import_set.go

```diff
@@ -0,0 +1,138 @@
+ */
+
+package astmodel
+
+import "sort"
+
+// PackageImportSet represents a set of distinct PackageImport references
+type PackageImportSet struct {
+	imports map[PackageImport]struct{}
+}
+
+// EmptyPackageImportSet creates a new empty set of PackageImport references
+func EmptyPackageImportSet() *PackageImportSet {
+	return &PackageImportSet{
+		imports: make(map[PackageImport]struct{}),
+	}
+}
+
+// AddImport ensures the set includes the specified import
+// Adding an import already present in the set is fine.
+func (set *PackageImportSet) AddImport(packageImport PackageImport) {
+	set.imports[packageImport] = struct{}{}
+}
+
+// AddReference ensures this set includes an import of the specified reference
+// Adding a reference already in the set is fine.
+func (set *PackageImportSet) AddReference(ref PackageReference) {
+	set.AddImport(NewPackageImport(ref))
+}
+
+// Merge ensures that all imports specified in other are included
+func (set *PackageImportSet) Merge(other *PackageImportSet) {
+	for i := range other.imports {
+		set.AddImport(i)
+	}
+}
+
+// Remove ensures the specified item is not present
+// Removing an item not in the set is not an error.
+func (set *PackageImportSet) Remove(packageImport PackageImport) {
+	delete(set.imports, packageImport)
+}
+
+// Contains allows checking to see if an import is included
+func (set *PackageImportSet) ContainsImport(packageImport PackageImport) bool {
+	_, found := set.imports[packageImport]
+	return found
+}
+
+// ImportFor looks up a package reference and returns its import, if any
+func (set *PackageImportSet) ImportFor(ref PackageReference) (PackageImport, bool) {
+	for imp := range set.imports {
+		if imp.PackageReference.Equals(ref) {
+			return imp, true
+		}
+	}
+
+	return PackageImport{}, false
+}
+
+// AsSlice() returns a slice containing all the imports
+func (set *PackageImportSet) AsSlice() []PackageImport {
+	var result []PackageImport
+	for i := range set.imports {
+		result = append(result, i)
+	}
+
+	sort.Slice(result, func(i, j int) bool {
+		return result[i].String() < result[j].String()
+	})
+
+	return result
+}
+
+// AsSortedSlice() return a sorted slice containing all the imports
+// less specifies how to order the imports
+func (set *PackageImportSet) AsSortedSlice(less func(i PackageImport, j PackageImport) bool) []PackageImport {
+	result := set.AsSlice()
+
+	sort.Slice(result, func(i int, j int) bool {
+		return less(result[i], result[j])
+	})
+
+	return result
+}
+
+// Length returns the number of unique imports in this set
+func (set *PackageImportSet) Length() int {
+	return len(set.imports)
+}
+
+// ApplyName replaces any existing PackageImport for the specified reference with one using the
+// specified name
+func (set *PackageImportSet) ApplyName(ref PackageReference, name string) {
+	found := false
+	// Iterate over a slice as that freezes the list
+	for _, imp := range set.AsSlice() {
+		if imp.PackageReference.Equals(ref) {
+			set.Remove(imp)
```

Comment:
There's a possible problem here (I'm not 100% sure it's a problem but it might be):
This set allows representation of multiple imports with the same package but different names. That's really an error (we have a TODO I believe to check for it and resolve it), but the way this function is written it will remove all N occurrences of an import with matching `PackageReference`s and replace them with a single one (the one so named).

On the one hand... this function almost could be used to solve the TODO - on the other hand it's not clear that is the intent of this function and it feels like that might be an unintentional side-effect rather than the real goal.

I wonder if it would be better to make the caller supply the specific `PackageImport` they want to modify - that way you can do a direct lookup and it also sidesteps the duplicates problem. I think doing this at the only current callsite would be easy.
Implementation could also then be something like the below (which seems a bit cleaner to me):
```
current, ok := set.imports[packageImport]
if !ok {
    return
}
set.Remove(imp)
set.AddImport(packageImport.WithName(name))
```

Reply:
The implementation was deliberate - and would only have impacted things if there were multiple imports for the same reference. I've changed the type definition to use `map[PackageReference]PackageImport` which makes the implementation here trivial and prevents multiple incompatible imports of the same reference, solving the TODO as well.